python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -921,54 +874,98 @@ def execute_asset_backfill_iteration(
+            result = None
+            previous_requested_subset = previous_asset_backfill_data.requested_subset
+
+            for result in execute_asset_backfill_iteration_inner(
+                backfill_id=backfill.backfill_id,
+                asset_backfill_data=previous_asset_backfill_data,
+                instance_queryer=instance_queryer,
+                asset_graph=asset_graph,
+                backfill_start_timestamp=backfill.backfill_timestamp,
+                logger=logger,
+            ):
+                yield None
 
-        if not isinstance(result, AssetBackfillIterationResult):
-            check.failed(
-                "Expected execute_asset_backfill_iteration_inner to return an"
-                " AssetBackfillIterationResult"
+            if not isinstance(result, AssetBackfillIterationResult):
+                check.failed(
+                    "Expected execute_asset_backfill_iteration_inner to return an"
+                    " AssetBackfillIterationResult"
+                )
+
+            requested_partitions = _get_requested_asset_partitions_from_run_requests(
+                result.run_requests,
+                asset_graph,
+                instance_queryer,
             )
 
-        updated_asset_backfill_data = result.backfill_data
+            requested_partitions_subset = (
+                previous_requested_subset
+                | AssetGraphSubset.from_asset_partition_set(
+                    set(requested_partitions), asset_graph=asset_graph
+                )
+            )
+
+            if requested_partitions_subset != result.backfill_data.requested_subset:
+                missing_partitions = list(
+                    (
+                        result.backfill_data.requested_subset - requested_partitions_subset
+                    ).iterate_asset_partitions()
+                )
+                check.failed(
+                    "Did not request runs for all expected partitions. \n\nPartitions not"
+                    f" requested: {missing_partitions}",
+                )
+
+            # Write the updated asset backfill data with in progress run requests before we launch anything, for idempotency
+            # Make sure we didn't get canceled in the interim
+            backfill = cast(PartitionBackfill, instance.get_backfill(backfill.backfill_id))
+            if backfill.status != BulkActionStatus.REQUESTED:
+                logger.info("Backfill was canceled mid-iteration, returning")
+                return
+
+            backfill = (
+                backfill.with_asset_backfill_data(
+                    result.backfill_data,
+                    dynamic_partitions_store=instance,
+                    asset_graph=asset_graph,
+                )
+                .with_in_progress_run_requests(result.run_requests, result.reserved_run_ids)
+                .with_failure_count(0)
+            )
+
+            instance.update_backfill(backfill)
```

Comment:
question for reviewers: one significant change here is that we now write the updated AssetBackfillData *before* we submit any runs (to ensure idempotence on retry), instead of after. Does that have any important effect on how the backfill is displayed in the UI during the period of time when the runs are being submitted?

Reply:
the problem with that is that we cannot rely on having the previous iteration's requested subset anymore (unless I add a field that stores it separately) - if this tick is a retry after we decided what run requests to launch but then one of them failed, we won't have access to it anymore

Do you think I should add that field separately and store it?